tests/integration/test_pip_packages.py

```diff
@@ -0,0 +1,116 @@
+    ref = test_env["pip_packages"]["without_deps"]["ref"]
+    pkg_managers = ["pip"]
+    initial_response = client.create_new_request(
+        payload={
+            "repo": repo,
+            "ref": ref,
+            "pkg_managers": pkg_managers,
+            # TODO: delete pip-dev-preview flag when
+            #  the pip package manager will be ready for production usage
+            "flags": ["pip-dev-preview"],
+        },
+    )
+    completed_response = client.wait_for_complete_request(initial_response)
+    assert completed_response.status == 200
+
+    assert_packages_from_response(completed_response.data, 1, pkg_managers)
+
+    assert_dependencies_from_cachito_request(completed_response.data)
+
+    # Download expected source files
+    expected_files_urls = test_env["pip_packages"]["without_deps"]["expected_files"]
+    expected_source_files = {}
+    for file_name, file_url in expected_files_urls.items():
+        expected_source_files[file_name] = get(file_url).text
+
+    # Download and extract source tarball
+    source_name = tmpdir.join(f"download_{str(completed_response.id)}")
+    file_name_tar = tmpdir.join(f"download_{str(completed_response.id)}.tar.gz")
+    resp = client.download_bundle(completed_response.id, file_name_tar)
+    assert resp.status == 200
+    assert tarfile.is_tarfile(file_name_tar)
+    with tarfile.open(file_name_tar, "r:gz") as tar:
+        tar.extractall(source_name)
+
+    # Check that the source tarball includes the application source code under the app directory.
+    assert_expected_files(path.join(source_name, "app"), expected_source_files)
+    # Check that the source tarball includes an empty deps directory.
+    assert_expected_files(path.join(source_name, "deps"))
+
+
+def assert_packages_from_response(response_data, pkg_count, pkg_managers):
+    """Check amount and type of packages in the response data."""
+    assert len(response_data["packages"]) == pkg_count
+    for idx in range(pkg_count):
+        assert response_data["packages"][idx]["type"] == pkg_managers[idx]
+
+
+def assert_dependencies_from_cachito_request(response_data, expected_deps=None):
+    """
+    Check that dependencies are correctly listed in the response data.
+
+    :param dict response_data: response data from the Cachito request
+    :param list expected_deps: expected package dependencies
+    """
+    if expected_deps is None:
+        expected_deps = []
+
+    assert set(response_data["dependencies"]) == set(expected_deps)
+
+    packages_deps = []
+    for package in response_data["packages"]:
+        packages_deps += package["dependencies"]
+    assert set(packages_deps) == set(expected_deps)
+
+
+def assert_expected_files(source_path, expected_files=None):
+    """
+    Check that the source path includes expected files.
+
+    :param str source_path: local path for checking
+    :param dict expected_files: {"relative_path/file_name: "file_content", ...}
```

Comment:
Should this function receive the URLs and open the files here in a context manager and do the comparison file by file instead to avoid loading all files in memory and passing them around?

Reply:
Done, thank you